docs/HKT.md

```diff
@@ -0,0 +1,168 @@
+}
+
+export function lift<F>(F: Functor<F>): <A, B>(f: (a: A) => B) => (fa: HKT<F, A>) => HKT<F, B> {
+  return f => fa => F.map(f, fa)
+}
+```
+
+We can define an instance of `Functor` for `Identity`
+
+```ts
+// Identity.ts
+
+export const URI = 'Identity'
+
+export type URI = typeof URI
+
+export class Identity<A> {
+  readonly _A: A
+  readonly _URI: URI
+  constructor(readonly value: A) {}
+  map<B>(f: (a: A) => B): Identity<B> {
+    return new Identity(f(this.value))
+  }
+}
+
+export const map = <A, B>(f: (a: A) => B, fa: Identity<A>): Identity<B> => {
+  return fa.map(f)
+}
+
+export const identity: Functor<URI> = {
+  URI,
+  map
+}
+```
+
+## The problem
+
+`Identity` behaves as expected when using its `map`
+
+```ts
+const double = (n: number): number => n * 2
+
+// x: Identity<number>
+const x = map(double, new Identity(1))
+```
+
+However there's a problem with functions which abstracts over `Functor` like `lift`
+
+```ts
+// liftedDouble: (fa: HKT<"Identity", number>) => HKT<"Identity", number>
+const liftedDouble = lift(identity)(double)
+
+// x: HKT<"Identity", number>
+const x = liftedDouble(new Identity(1))
+```
+
+`x` is not usable
+
+```ts
+x.value // static error: Property 'value' does not exist on type 'HKT<"Identity", number>'
+```
+
+## The solution
+
+We must somehow teach TypeScript that `HKT<"Identity", number>` is really `Identity<number>`, or more generally that `HKT<"Identity", A>` is `Identity<number>` for all `A`.
+
+### First step: build a type level map `URI -> Type constructor`
+
+The type level map is named `URI2HKT`
+
+```ts
+// HKT.ts
+
+export interface URI2HKT<A> {}
+```
+
+Adding an entry means to leverage the module augmentation feature
+
+```ts
+// Identity.ts
+
+declare module './HKT' {
+  interface URI2HKT<A> {
+    Identity: Identity<A> // maps the URI "Identity" to the type constructor `Identity`
+  }
+}
+```
+
+### Second step: add a specialized overloading to `lift`
+
+If `F` is an `URI` which corresponds to a key in `URI2HKT` then we can add a specialized overloading for it to `lift`
+
+```ts
+// specialized overloading
+export function lift<F extends keyof URI2HKT<any>>(
+  F: Functor<F>
+): <A, B>(f: (a: A) => B) => (fa: URI2HKT<A>[F]) => URI2HKT<B>[F]
+ // keep the generic signature
+export function lift<F>(F: Functor<F>): <A, B>(f: (a: A) => B) => (fa: HKT<F, A>) => HKT<F, B>
+export function lift<F>(F: Functor<F>): <A, B>(f: (a: A) => B) => (fa: HKT<F, A>) => HKT<F, B> {
+  return f => fa => F.map(f, fa)
+}
+```
+
+As soon as we add the specialized overloading we get the desired behavior
+
+```ts
+// liftedDouble: (fa: Identity<number>) => Identity<number>
+const liftedDouble = lift(identity)(double)
+
+// x: Identity<number>
+const x = liftedDouble(new Identity(1))
+
+x.value // ok
+```
+
+In order to make easier to define an overloading, we can write some type aliases
+
+```ts
+// HKT.ts
+
+export type HKTS = keyof URI2HKT<any>
+
+export type HKTAs<F extends HKTS, A> = URI2HKT<A>[F]
```

Comment:
And `HKTAs` extracts actual registered type constructor from the storage by its URI and also for passed A type argument. So it's a final usable type.